File: desktop-client/src/app/game/services/submit-manager/submit-manager.service.ts
import { Injectable } from '@angular/core';
import { KonvaService } from '@app/game/services/konva/konva.service';
import { RoomService } from '@app/home/services/room/room.service';
import { S3Service } from '@app/shared/services/s3/s3.service';
import { QuestionType } from '@common/enum/question-type';
import { PlayerAnswer } from '@common/interfaces/player-answer';
import { Question } from '@common/interfaces/question';
@Injectable({
    providedIn: 'root',
})
export class SubmitManagerService {
    buttonsAreActive: boolean[] = [];
    disabledAnswerChoices: Set<number> = new Set();
    textAnswer: string = '';
    numericalAnswer: number;

    constructor(
        private readonly konvaService: KonvaService,
        private readonly s3Service: S3Service,
        private readonly roomService: RoomService,
    ) {}

    setNumberOfButtons(numButtons: number) {
        this.buttonsAreActive = Array(numButtons).fill(false);
    }

    buttonWasClicked(index: number) {
        this.buttonsAreActive[index] = !this.buttonsAreActive[index];
        this.buttonsAreActive = [...this.buttonsAreActive];
    }

    changeTextAnswer(answer: string) {
        this.textAnswer = answer;
    }

    changeNumericalAnswer(answer: number) {
        this.numericalAnswer = answer;
    }

    setInputs(question: Question) {
        this.textAnswer = '';

        if (question.upperBound !== undefined && question.lowerBound !== undefined) {
            this.numericalAnswer = Math.round((question.upperBound + question.lowerBound) / 2);
        } else {
            this.numericalAnswer = 0;
        }

        if (question.type === QuestionType.MultiChoice) {
            this.disabledAnswerChoices.clear();
            this.buttonsAreActive = new Array(question.choices.length).fill(false);
        }
    }

    answerIsChosen(questionType: QuestionType): boolean {
        switch (questionType) {
            case QuestionType.MultiChoice:
                return this.buttonsAreActive.includes(true);
            case QuestionType.OpenEnded:
                return this.textAnswer.length > 0;
            case QuestionType.Drawing:
            case QuestionType.Estimation:
                return true;
        }
    }

    async getAnswer(questionType: QuestionType): Promise<PlayerAnswer> {
        let awsKey = '';
        if (questionType === QuestionType.Drawing) {
            const blob = (await this.konvaService.getDrawingBlob()) ?? ({} as unknown as Blob);
            awsKey = `drawings/${this.roomService.roomId}/${crypto.randomUUID()}.png`;
            await this.s3Service.uploadBlobImage(blob, awsKey);
        }
        switch (questionType) {
            case QuestionType.MultiChoice:
                return { questionType, data: this.buttonsAreActive };
            case QuestionType.OpenEnded:
                return { questionType, data: this.textAnswer };
            case QuestionType.Estimation:
                return { questionType, data: this.numericalAnswer };
            case QuestionType.Drawing:
                return { questionType, data: awsKey };
        }
    }
}

File: mobile_client/lib/services/game/submit_manager_service.dart
import 'dart:typed_data';

import 'package:flutter/material.dart';
import 'package:mobile_client/constants/enum/question_type.dart';
import 'package:mobile_client/models/game/player_answer.dart';
import 'package:mobile_client/models/question.dart';
import 'package:mobile_client/services/flutter_painter_service.dart';
import 'package:mobile_client/services/room_service.dart';
import 'package:mobile_client/services/s3_service.dart';
import 'package:mobile_client/utils/void_notifier.dart';
import 'package:uuid/uuid.dart';

class SubmitManagerService extends ChangeNotifier {
  final FlutterPainterService _flutterPainterService;
  final S3Service _s3service;
  final RoomService _roomService;
  final Uuid uuid = Uuid();

  SubmitManagerService(
      this._flutterPainterService, this._s3service, this._roomService);

  List<bool> buttonsAreActive = [];
  Set<int> disabledAnswerChoices = {};

  String _textAnswer = '';
  int numericalAnswer = 0;
  VoidNotifier resetInputs = VoidNotifier();

  void buttonWasClicked(int index) {
    buttonsAreActive[index] = !buttonsAreActive[index];
    buttonsAreActive = [...buttonsAreActive];
    notifyListeners();
  }

  void changeTextAnswer(String answer) {
    _textAnswer = answer;
    notifyListeners();
  }

  void changeNumericalAnswer(int answer) {
    numericalAnswer = answer;
    notifyListeners();
  }

  void setInputs(Question question) {
    _textAnswer = '';
    numericalAnswer = 0;

    resetInputs.notify();
    if (question.type == QuestionType.multiChoice) {
      disabledAnswerChoices.clear();
      buttonsAreActive = List.filled(question.choices.length, false);
    }
    notifyListeners();
  }

  bool answerIsChosen(QuestionType questionType) {
    switch (questionType) {
      case QuestionType.multiChoice:
        return buttonsAreActive.contains(true);
      case QuestionType.openEnded:
        return _textAnswer.isNotEmpty;
      case QuestionType.drawing:
      case QuestionType.estimation:
        return true;
    }
  }

  Future<PlayerAnswer> getAnswer(QuestionType questionType) async {
    switch (questionType) {
      case QuestionType.multiChoice:
        return MultiChoicePlayerAnswer(data: buttonsAreActive);
      case QuestionType.openEnded:
        return OpenEndedPlayerAnswer(data: _textAnswer);
      case QuestionType.estimation:
        return EstimationPlayerAnswer(data: numericalAnswer);
      case QuestionType.drawing:
        Uint8List? blob = await _flutterPainterService.getDrawingBlob();
        String awsKey = "drawings/${_roomService.roomId}/${uuid.v4()}.png";
        await _s3service.uploadBlobImage(blob ?? Uint8List(0), awsKey);
        return DrawingPlayerAnswer(data: awsKey);
    }
  }
}
